ic-factory/src/core.rs

```diff
@@ -1,124 +1,37 @@
-    pub fn len(&self) -> usize {
-        self.canisters.len()
-    }
-
-    /// Returns true if no canister is created yet.
-    pub fn is_empty(&self) -> bool {
-        self.canisters.is_empty()
-    }
-
-    /// Returns a vector of all ic-helpers created by the factory.
-    pub fn all(&self) -> Vec<Principal> {
-        self.canisters
-            .values()
-            .map(|canister| canister.identity())
-            .collect()
-    }
-
-    /// Creates a pair with cycles in it to make it workable.
-    ///
-    /// The amount of cycles that will be available in the created canister is `cycles - FEE`, where
-    /// `FEE` is a constant value needed to cover the factory expenses. Current implementation has
-    /// `FEE == 10^11 + 10^6`.
-    pub fn create_with_cycles<A: ArgumentEncoder>(
-        &self,
-        wasm_module: &[u8],
-        arg: A,
-        cycles: u64,
-    ) -> impl Future<Output = CallResult<Canister>> {
-        Canister::create(self.checksum.version, wasm_module.into(), arg, cycles)
-    }
-
-    /// Stops and deletes the canister. After this actor is awaited on, [Factory::forget] method must be used
-    /// to remove the canister from the list of created canisters.
-    pub fn drop(&self, canister: Principal) -> impl Future<Output = Result<(), FactoryError>> {
-        drop_canister(canister)
-    }
-
-    /// Adds a new canister to the canister registry. If a canister with the given key is already
-    /// registered, it will be replaced with the new one.
-    pub fn register(&mut self, key: Principal, canister: Canister) {
-        self.canisters.insert(key, canister);
-    }
-
-    /// Removes the canister from the registry. Return error if the canister with the given key is
-    /// not registered.
-    pub fn forget(&mut self, key: &Principal) -> Result<(), FactoryError> {
-        self.canisters
-            .remove(key)
-            .ok_or(FactoryError::NotFound)
-            .map(|_| ())
-    }
-
-    /// Returns a future that upgrades a canister to the given bytecode. After the future is
-    /// done executing, `register_upgraded` method shall be called to add the resulting canister to the
-    /// registry.
-    ///
-    /// Please, note that the state should not be borrowed when this future is awaited on, to prevent
-    /// memory access conflict in case of concurrent requests.
-    pub fn upgrade(
-        &self,
-        canister: &Canister,
-        wasm_module: Vec<u8>,
-    ) -> impl Future<Output = CallResult<Canister>> {
-        upgrade_canister(self.checksum.version, canister.clone(), wasm_module)
-    }
+pub async fn create_canister<T: ArgumentEncoder>(
+    wasm_module: Vec<u8>,
+    init_args: T,
+    cycles: u64,
+    controllers: Option<Vec<Principal>>,
+) -> CallResult<Principal> {
+    let settings = CanisterSettings {
+        controllers,
+        compute_allocation: None,
+        memory_allocation: None,
+        freezing_threshold: None,
```

Comment:
`CanisterSettings` implements `Default`
```suggestion
        ..Default::default()
```